### DISCUSSION TITLE
Help with debug

### DISCUSSION THREAD

**Discussion by sylman4** [apprenticeharper/DeDRM_tools#1989] [Q&A]
I'm really new at this but I'm trying to dedrm a textbook I've downloaded for school through the Kindle App. I would really appreciate it if you could look it over for me and figure out what I'm doing wrong. Thanks! I've pasted the debug code below:

calibre Debug log
calibre 5.34  embedded-python: True is64bit: True
macOS-10.16-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '21.1.0', 'Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:01 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T6000')
Python 3.8.5
OSX: ('10.16', ('', '', ''), 'x86_64')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 2, 1)
calibre 5.34  embedded-python: True is64bit: True
macOS-10.16-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '21.1.0', 'Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:01 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T6000')
Python 3.8.5
OSX: ('10.16', ('', '', ''), 'x86_64')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 2, 1)
devicePixelRatio: 2.0
logicalDpi: 72.0 x 72.0
physicalDpi: 127.50000191485789 x 127.50000191485789
Using calibre Qt style: True
[0.00] Starting up...
[0.02] Showing splash screen...
[0.22] splash screen shown
[0.22] Initializing db...
[0.23] db initialized
[0.23] Constructing main UI...
[1.02] main UI initialized...
[1.02] Hiding splash screen
Starting QuickView
DeDRM v7.2.1: Trying to decrypt B08MVC8359_EBOK.azw
Using Library AlfCrypto Python
Using Library AlfCrypto Python
DeDRM v7.2.1: Failed to decrypt with error: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.
DeDRM v7.2.1: Looking for new default Kindle Key after 0.1 seconds
Found k4Mac kinf2018 file: /Users/lisawan/Library/Containers/com.amazon.Kindle/Data/Library/Application Support/Kindle/storage/.kinf2018
trying username  b'lisawan'  on file  /Users/l/Library/Containers/com.amazon.Kindle/Data/Library/Application Support/Kindle/storage/.kinf2018
trying IDString: b'c31e401bc16d'
trying IDString: b'c31e401bc16c'
[3.76] splash screen hidden
[3.76] Started up in 3.76 seconds with 1 books
trying IDString: b'c31e401bc16f'
trying IDString: b'5de82cf4c5c8'
Decrypted key file using IDString '5de82cf4c5c8' and UserName 'lisawan'
DeDRM v7.2.1: Found 1 new key
DeDRM v7.2.1: Ultimately failed to decrypt after 2.0 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 540, in KindleMobiDecrypt
    book = k4mobidedrm.GetDecryptedBook(path_to_ebook,kindleDatabases,androidFiles,serials,pids,self.starttime)
  File "calibre_plugins.dedrm.k4mobidedrm", line 208, in GetDecryptedBook
    raise DrmException("The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.")
calibre_plugins.dedrm.k4mobidedrm.DrmException: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre/customize/ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 633, in run
    decrypted_ebook = self.KindleMobiDecrypt(path_to_ebook)
  File "calibre_plugins.dedrm.__init__", line 583, in KindleMobiDecrypt
    raise DeDRMError("{0} v{1}: Ultimately failed to decrypt after {2:.1f} seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md".format(PLUGIN_NAME, PLUGIN_VERSION,time.time()-self.starttime))
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 2.0 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Traceback (most recent call last):
  File "calibre/customize/ui.py", line 432, in get_file_type_metadata
  File "calibre/customize/builtins.py", line 260, in get_metadata
  File "calibre/ebooks/metadata/mobi.py", line 504, in get_metadata
  File "calibre/ebooks/mobi/reader/headers.py", line 294, in __init__
  File "calibre/ebooks/mobi/reader/headers.py", line 322, in identity
calibre.ebooks.mobi.MobiError: Unknown book type: b'\x02\xa2\x81\xc4\xde\x02\x9d\x8c'

**Comment by ElleKayEm** [comment]
First, I recommend using [noDRM's DeDRM](https://github.com/noDRM/DeDRM_tools/releases) because it has some Kindle for Mac related improvements.  Then I recommend avoiding KFX format if possible.  Take a look at method 2 in this post:  https://www.mobileread.com/forums/showthread.php?t=283371

And also this post:  https://www.mobileread.com/forums/showpost.php?p=3819708&postcount=508

**Reply by sylman4** [reply]
ok got it. I think I'm getting somewhere with the KFX but I'm still having trouble with the DRM. Here's what I have now:

calibre Debug log
calibre 5.34  embedded-python: True is64bit: True
macOS-10.16-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '21.1.0', 'Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:01 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T6000')
Python 3.8.5
OSX: ('10.16', ('', '', ''), 'x86_64')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 2, 1)
calibre 5.34  embedded-python: True is64bit: True
macOS-10.16-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '21.1.0', 'Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:01 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T6000')
Python 3.8.5
OSX: ('10.16', ('', '', ''), 'x86_64')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 2, 1)
devicePixelRatio: 2.0
logicalDpi: 72.0 x 72.0
physicalDpi: 127.50000191485789 x 127.50000191485789
Using calibre Qt style: True
[0.00] Starting up...
[0.02] Showing splash screen...
[0.23] splash screen shown
[0.23] Initializing db...
[0.23] db initialized
[0.23] Constructing main UI...
[0.88] main UI initialized...
[0.88] Hiding splash screen
Starting QuickView
DeDRM v7.2.1: Trying to decrypt B08MVC8359_EBOK.azw
Using Library AlfCrypto Python
Using Library AlfCrypto Python
MobiDeDrm v1.0.
Copyright © 2008-2020 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: The Complete PCOA Review
Found 2 keys to try after 0.1 seconds
Crypto Type is: 2
DeDRM v7.2.1: Failed to decrypt with error: No key found in 2 PIDs tried.
DeDRM v7.2.1: Looking for new default Kindle Key after 0.1 seconds
Found k4Mac kinf2018 file: /Users/l/Library/Containers/com.amazon.Kindle/Data/Library/Application Support/Kindle/storage/.kinf2018
Found k4Mac kinf2018 file: /Users/l/Library/Application Support/Kindle/storage/.kinf2018
trying username  b'l'  on file  /Users/l/Library/Containers/com.amazon.Kindle/Data/Library/Application Support/Kindle/storage/.kinf2018
trying IDString: b'c31e401bc16d'
trying IDString: b'c31e401bc16c'
trying IDString: b'c31e401bc16f'
trying IDString: b'5de82cf4c5c8'
[2.65] splash screen hidden
[2.65] Started up in 2.65 seconds with 0 books
Decrypted key file using IDString '5de82cf4c5c8' and UserName 'l'
trying username  b'l'  on file  /Users/l/Library/Application Support/Kindle/storage/.kinf2018
trying IDString: b'c31e401bc16d'
trying IDString: b'c31e401bc16c'
trying IDString: b'c31e401bc16f'
trying IDString: b'5de82cf4c5c8'
Decrypted key file using IDString '5de82cf4c5c8' and UserName 'l'
DeDRM v7.2.1: Found 2 new keys
MobiDeDrm v1.0.
Copyright © 2008-2020 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: The Complete PCOA Review
Got SerialNumber from database default_key_1
Got UsernameHash from database default_key_1
Got SerialNumber from database default_key_2
Got UsernameHash from database default_key_2
Found 8 keys to try after 4.0 seconds
Crypto Type is: 2
DeDRM v7.2.1: Ultimately failed to decrypt after 4.0 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 540, in KindleMobiDecrypt
    book = k4mobidedrm.GetDecryptedBook(path_to_ebook,kindleDatabases,androidFiles,serials,pids,self.starttime)
  File "calibre_plugins.dedrm.k4mobidedrm", line 238, in GetDecryptedBook
    mb.processBook(totalpids)
  File "calibre_plugins.dedrm.mobidedrm", line 484, in processBook
    raise DrmException("No key found in {0:d} PIDs tried.".format(len(goodpids)))
calibre_plugins.dedrm.mobidedrm.DrmException: No key found in 2 PIDs tried.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre/customize/ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 633, in run
    decrypted_ebook = self.KindleMobiDecrypt(path_to_ebook)
  File "calibre_plugins.dedrm.__init__", line 583, in KindleMobiDecrypt
    raise DeDRMError("{0} v{1}: Ultimately failed to decrypt after {2:.1f} seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md".format(PLUGIN_NAME, PLUGIN_VERSION,time.time()-self.starttime))
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 4.0 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md

**Comment by ZolaLa9** [comment]
It looks like you're still using the older DeDRM v7.2.1. This version has a typo in the Mac key retrieval script. 

Try the latest [DeDRM v10.0.2](https://github.com/noDRM/DeDRM_tools/releases/tag/v10.0.2).
